File: test/clients/go/persistent_ddl_dml.go
// Command persistent_ddl_dml is the official cloud.google.com/go/bigquery
// smoke test for persistent SQL mutations and streaming inserts, mirroring
// test/clients/python/persistent_ddl_dml.py and
// test/clients/node/persistent_ddl_dml.js so all three official clients
// exercise the same real server behavior.
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"log"

	"cloud.google.com/go/bigquery"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
)

func runQuery(ctx context.Context, client *bigquery.Client, sql string) *bigquery.JobStatus {
	q := client.Query(sql)
	job, err := q.Run(ctx)
	if err != nil {
		log.Fatalf("run query %q: %v", sql, err)
	}
	status, err := job.Wait(ctx)
	if err != nil {
		log.Fatalf("wait for job %q: %v", sql, err)
	}
	if status.Err() != nil {
		log.Fatalf("job for %q failed: %v", sql, status.Err())
	}
	return status
}

func queryStats(status *bigquery.JobStatus) *bigquery.QueryStatistics {
	stats, ok := status.Statistics.Details.(*bigquery.QueryStatistics)
	if !ok {
		log.Fatalf("expected *bigquery.QueryStatistics, got %T", status.Statistics.Details)
	}
	return stats
}

func runDML(ctx context.Context, client *bigquery.Client, sql string, expectedAffectedRows int64) *bigquery.QueryStatistics {
	stats := queryStats(runQuery(ctx, client, sql))
	if stats.NumDMLAffectedRows != expectedAffectedRows {
		log.Fatalf("expected %d affected rows for %q, got %d", expectedAffectedRows, sql, stats.NumDMLAffectedRows)
	}
	return stats
}

type mutationRow struct {
	ID     int64               `bigquery:"id"`
	Name   bigquery.NullString `bigquery:"name"`
	Active bool                `bigquery:"active"`
}

func main() {
	endpoint := flag.String("endpoint", "http://127.0.0.1:19050", "LocaQL REST endpoint")
	flag.Parse()

	ctx := context.Background()
	// The client's generated REST stubs derive their full request path from
	// this endpoint directly rather than appending /bigquery/v2 themselves,
	// unlike the endpoint form the Python/Node clients expect (host only) —
	// confirmed empirically: a bare host produced 404s on LocaQL's own
	// request logs for /projects/p1/jobs (missing /bigquery/v2 entirely).
	client, err := bigquery.NewClient(ctx, "p1",
		option.WithEndpoint(*endpoint+"/bigquery/v2/"),
		option.WithoutAuthentication(),
	)
	if err != nil {
		log.Fatalf("new client: %v", err)
	}
	defer client.Close()

	runQuery(ctx, client, "CREATE OR REPLACE TABLE analytics.go_sql_mutations (id INT64, name STRING, active BOOL)")

	inserted := runDML(ctx, client,
		"INSERT INTO analytics.go_sql_mutations (id, name, active) VALUES (1, 'one', TRUE), (2, NULL, FALSE)", 2)
	updated := runDML(ctx, client,
		"UPDATE analytics.go_sql_mutations SET name = 'two', active = TRUE WHERE id = 2", 1)

	runQuery(ctx, client,
		"CREATE OR REPLACE TABLE analytics.go_sql_merge_source AS "+
			"SELECT 2 AS id, 'two-merged' AS name, FALSE AS active UNION ALL SELECT 3, 'three', TRUE")
	merged := runDML(ctx, client,
		"MERGE INTO analytics.go_sql_mutations AS T "+
			"USING analytics.go_sql_merge_source AS S ON T.id = S.id "+
			"WHEN MATCHED THEN UPDATE SET name = S.name, active = S.active "+
			"WHEN NOT MATCHED THEN INSERT (id, name, active) VALUES (S.id, S.name, S.active)", 2)
	deleted := runDML(ctx, client, "DELETE FROM analytics.go_sql_mutations WHERE id = 1", 1)

	// Streaming insert (tabledata.insertAll) via the official Inserter, not a
	// query job — real BigQuery Go client code uses this for row ingestion.
	inserter := client.Dataset("analytics").Table("go_sql_mutations").Inserter()
	if err := inserter.Put(ctx, []*mutationRow{{ID: 4, Name: bigquery.NullString{StringVal: "four", Valid: true}, Active: true}}); err != nil {
		log.Fatalf("streaming insert: %v", err)
	}

	it, err := client.Query("SELECT id, name, active FROM analytics.go_sql_mutations ORDER BY id").Read(ctx)
	if err != nil {
		log.Fatalf("read rows: %v", err)
	}
	type row struct {
		ID     int64
		Name   string
		Active bool
	}
	var got []row
	for {
		var r row
		err := it.Next(&r)
		if err == iterator.Done {
			break
		}
		if err != nil {
			log.Fatalf("iterate rows: %v", err)
		}
		got = append(got, r)
	}
	want := []row{
		{ID: 2, Name: "two-merged", Active: false},
		{ID: 3, Name: "three", Active: true},
		{ID: 4, Name: "four", Active: true},
	}
	if len(got) != len(want) {
		log.Fatalf("persistent DML + streaming insert rows mismatch: got %+v, want %+v", got, want)
	}
	for i := range got {
		if got[i] != want[i] {
			log.Fatalf("persistent DML + streaming insert rows mismatch at %d: got %+v, want %+v", i, got[i], want[i])
		}
	}

	runQuery(ctx, client, "DROP TABLE analytics.go_sql_merge_source")

	rowsOut := make([]map[string]any, len(got))
	for i, r := range got {
		rowsOut[i] = map[string]any{"id": r.ID, "name": r.Name, "active": r.Active}
	}
	out, err := json.Marshal(map[string]any{
		"delete": deleted.NumDMLAffectedRows,
		"insert": inserted.NumDMLAffectedRows,
		"merge":  merged.NumDMLAffectedRows,
		"rows":   rowsOut,
		"update": updated.NumDMLAffectedRows,
	})
	if err != nil {
		log.Fatalf("marshal report: %v", err)
	}
	fmt.Println(string(out))
}

File: test/clients/php/persistent_ddl_dml.php
<?php
/**
 * Official google/cloud-bigquery smoke test for persistent SQL mutations and
 * streaming inserts, mirroring
 * test/clients/{python,node,go,java,ruby}/persistent_ddl_dml.* so every
 * official client exercises the same real server behavior.
 *
 * The official PHP client has no documented anonymous-credentials mode for a
 * custom local endpoint (unlike the Go/Java clients' AnonymousCredentials/
 * NoCredentials). BigQueryClient's constructor accepts a `credentialsFetcher`
 * option typed `Google\Auth\FetchAuthTokenInterface` — confirmed by reading
 * the installed source (vendor/google/cloud-bigquery/src/BigQueryClient.php),
 * not documentation — which is a plain 3-method interface with no real OAuth
 * machinery behind it. A fetcher returning a literal, never-refreshed fake
 * token satisfies it without touching real credential material, the same
 * "fake bearer token nobody validates" approach already used for dbt/bq.
 */

require __DIR__ . '/vendor/autoload.php';

use Google\Auth\FetchAuthTokenInterface;
use Google\Cloud\BigQuery\BigQueryClient;

final class FakeTokenCredentials implements FetchAuthTokenInterface
{
    public function fetchAuthToken(?callable $httpHandler = null)
    {
        return ['access_token' => 'locaql-fake-token', 'expires_in' => 3600];
    }

    public function getCacheKey()
    {
        return 'locaql-fake-token';
    }

    public function getLastReceivedToken()
    {
        return ['access_token' => 'locaql-fake-token', 'expires_at' => time() + 3600];
    }
}

function parse_endpoint(array $argv): string
{
    $endpoint = 'http://127.0.0.1:19050';
    foreach ($argv as $arg) {
        if (str_starts_with($arg, '--endpoint=')) {
            $endpoint = substr($arg, strlen('--endpoint='));
        }
    }
    return $endpoint;
}

function new_client(string $endpoint): BigQueryClient
{
    // RestTrait::getApiEndpoint only defaults to the https:// scheme when
    // the given value contains no "//" at all (confirmed reading the
    // installed source) — passing the endpoint through with its scheme
    // intact, rather than stripping it, is what keeps this on plain HTTP.
    return new BigQueryClient([
        'projectId' => 'p1',
        'apiEndpoint' => $endpoint,
        'credentialsFetcher' => new FakeTokenCredentials(),
    ]);
}

function run_query(BigQueryClient $bigQuery, string $sql)
{
    $results = $bigQuery->runQuery($bigQuery->query($sql));
    $job = $results->job();
    $job->reload();
    $info = $job->info();
    if (($info['status']['state'] ?? null) !== 'DONE') {
        throw new RuntimeException("query did not complete: $sql");
    }
    if (isset($info['status']['errorResult'])) {
        throw new RuntimeException("query failed: $sql: " . json_encode($info['status']['errorResult']));
    }
    return $info;
}

function run_dml(BigQueryClient $bigQuery, string $sql, int $expectedAffectedRows): int
{
    $info = run_query($bigQuery, $sql);
    $stats = $info['statistics']['query'] ?? [];
    $affected = (int) ($stats['numDmlAffectedRows'] ?? -1);
    if ($affected !== $expectedAffectedRows) {
        throw new RuntimeException(
            "expected $expectedAffectedRows affected rows for $sql, got $affected"
        );
    }
    $expectedStatementType = strtoupper(strtok(trim($sql), " \t\n"));
    $statementType = $stats['statementType'] ?? null;
    if ($statementType !== $expectedStatementType) {
        throw new RuntimeException(
            "expected statementType=$expectedStatementType, got " . var_export($statementType, true)
        );
    }
    return $affected;
}

function main(array $argv): void
{
    $bigQuery = new_client(parse_endpoint($argv));

    run_query($bigQuery, 'CREATE OR REPLACE TABLE analytics.php_sql_mutations (id INT64, name STRING, active BOOL)');

    $inserted = run_dml(
        $bigQuery,
        "INSERT INTO analytics.php_sql_mutations (id, name, active) VALUES (1, 'one', TRUE), (2, NULL, FALSE)",
        2
    );
    $updated = run_dml(
        $bigQuery,
        "UPDATE analytics.php_sql_mutations SET name = 'two', active = TRUE WHERE id = 2",
        1
    );

    run_query(
        $bigQuery,
        'CREATE OR REPLACE TABLE analytics.php_sql_merge_source AS ' .
        "SELECT 2 AS id, 'two-merged' AS name, FALSE AS active UNION ALL SELECT 3, 'three', TRUE"
    );
    $merged = run_dml(
        $bigQuery,
        'MERGE INTO analytics.php_sql_mutations AS T ' .
        'USING analytics.php_sql_merge_source AS S ON T.id = S.id ' .
        'WHEN MATCHED THEN UPDATE SET name = S.name, active = S.active ' .
        'WHEN NOT MATCHED THEN INSERT (id, name, active) VALUES (S.id, S.name, S.active)',
        2
    );
    $deleted = run_dml($bigQuery, 'DELETE FROM analytics.php_sql_mutations WHERE id = 1', 1);

    // Streaming insert (tabledata.insertAll) through the same table
    // resource, not a query job.
    $table = $bigQuery->dataset('analytics')->table('php_sql_mutations');
    $insertResponse = $table->insertRow(['id' => 4, 'name' => 'four', 'active' => true]);
    if (!$insertResponse->isSuccessful()) {
        throw new RuntimeException('streaming insert failed: ' . json_encode($insertResponse->failedRows()));
    }

    $results = $bigQuery->runQuery(
        $bigQuery->query('SELECT id, name, active FROM analytics.php_sql_mutations ORDER BY id')
    );
    $got = [];
    foreach ($results as $row) {
        $got[] = [$row['id'], $row['name'], $row['active']];
    }
    $want = [[2, 'two-merged', false], [3, 'three', true], [4, 'four', true]];
    if ($got !== $want) {
        throw new RuntimeException('persistent DML + streaming insert rows mismatch: ' . json_encode($got));
    }

    run_query($bigQuery, 'DROP TABLE analytics.php_sql_merge_source');

    echo json_encode([
        'delete' => $deleted,
        'insert' => $inserted,
        'merge' => $merged,
        'rows' => $got,
        'update' => $updated,
    ]) . "\n";
}

main($argv);
